### ISSUE TITLE
Possible Emscripten bug - switch statements

### ISSUE THREAD

**Issue by kitCore** [emscripten-core/emscripten#9327]
Emscripten has a weird bug, that appears under the following circumstances using switch statements. Let's say you have a simple code containing a switch statement:

```
int main()
{
  int i = 0;
  if (i & 0x02000) return 0;
  switch (i)
  {
  case 0:
  case 2049:
    return 0;
  }
}
```

The bug appears, if the condition `|a-b| < 1024*n+n` is fulfilled, where n is the number of cases, a is the biggest integer value of all cases and b is the smallest integer value of all cases. If the condition is fulfilled, the size of the generated .wasm file scales with the difference of a and b. For two cases, the maximum difference for still fulfilling the condition is 2049. The generated .wasm file then has 2836 kB. If the difference of a and b is 2050 or greater, the bug does not appear and the .wasm file has 820 kB. 

Not only does this introduce performance issues when using WebAssembly.compile in a browser, but additionally the Chromium browser crashes on an Arm64 platform, if eight or more cases are used and the biggest difference is greater than 8192. The crash is caused by the if statement (see the example code) and is triggered by a function "is_int14(...)" in the chromium code.

I am using the Emscripten version 1.38.16 and create the .wasm files with the command:

> em++ -g -s WASM=1 -s -s SIDE_MODULE=1 -o test.wasm test.cpp

**Comment by kitCore** [comment]
Thank you kripken for your fast response!
I followed your advice and tried it with the latest-upstream. I now have the version 1.38.42. 

Apparently tables are indeed handled differently, as you mentioned. The size of the generated file changed slightly. With the example code of my first post I now have a .wasm file with 2767 kB instead of 2836 kB. Increasing the argument of the second case from 2049 to 2050 (and thereby avoiding the bug) I now have a .wasm file with 751 kB instead 820 kB. 

In conclusion, the bug still appears unfortunately...

**Comment by kripken** [comment]
Odd, I see different results. Is there any chance you were still using fastcomp? What does emcc -v say?

Here is what I get:
```
$ ./emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 1.38.42
clang version 10.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project edfaee08115376467d1c95573de991aebfdaeb42)
[..]
$ ./emcc a.cpp -s SIDE_MODULE -g
emcc:WARNING: output suffix .js requested, but wasm side modules are just wasm files; emitting only a .wasm, no .js
emscripten:WARNING: Wasm source map won't be usable in a browser without --source-map-base
$ ls -al a.out.wasm
-rw-r--r-- 1 azakai primarygroup 1038 Aug 29 14:31 a.out.wasm
```

**Comment by kitCore** [comment]
Thank you, kripken! 
You were right, I was somehow still using fastcomp. Nevertheless, using the latest-upstream solved the problem.
